### PR TITLE
Cleanup of lower_utils.cpp stage 1: Isolate out GpuLower usage

### PR DESCRIPTION
**Pull request by csarofeen**
Also remove `ir_utils::caMapExactConcreteId`

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/lower_utils.cpp [hunk 1]
```diff
@@ -235,35 +235,6 @@ bool isScalarOp(const Expr* expr) {
   return true;
 }
 
-bool hasBlockSync(const Expr* expr, const ThreadPredicateMap& pred_map) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.cpp:238]: Just moved.

File: torch/csrc/jit/codegen/cuda/lower_utils.cpp [hunk 2]
```diff
@@ -734,15 +585,154 @@ std::vector<Expr*> replaceInputsInExpr(
...
+      });
+
+  return all_swizzles;
+}
+
+} // namespace ir_utils
+
+namespace lower_utils {
+
+bool hasBlockSync(const Expr* expr, const ThreadPredicateMap& pred_map) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.cpp:612]: Just moved.

File: torch/csrc/jit/codegen/cuda/lower_utils.h [hunk 3]
```diff
@@ -90,8 +90,6 @@ TORCH_CUDA_CU_API TensorView* getTvOutput(const Expr*);
 // Returns the first input of Expr that is a TensorView
 TORCH_CUDA_CU_API TensorView* getTvInput(const Expr*);
 
-bool hasBlockSync(const Expr* expr, const ThreadPredicateMap& pred_map);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.h:93]: Just moved.

File: torch/csrc/jit/codegen/cuda/lower_utils.h [hunk 4]
```diff
@@ -272,6 +223,52 @@ struct TORCH_CUDA_CU_API IterDomainDependencySorter {
   const std::shared_ptr<const ComputeAtMap> compute_at_map_;
 };
 
+} // namespace ir_utils
+
+namespace lower_utils {
+
+bool hasBlockSync(const Expr* expr, const ThreadPredicateMap& pred_map);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.h:230]: Just moved.

File: torch/csrc/jit/codegen/cuda/lower_utils.cpp [hunk 5]
```diff
@@ -400,20 +371,6 @@ bool isGlobalLoadInit(const Expr* expr) {
   return false;
 }
 
-kir::Allocate* allocGlobalBufferForGridComm(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.cpp:403]: Just moved.

File: torch/csrc/jit/codegen/cuda/lower_utils.cpp [hunk 6]
```diff
@@ -734,15 +585,154 @@ std::vector<Expr*> replaceInputsInExpr(
...
+  } else if (expr->isA<BroadcastOp>()) {
+    const ParallelTypeBitmap pt_map =
+        GpuLower::current()->threadPredMap().getParallelBroadcastDomains(tv);
+    return pt_map.any();
+  }
+
+  return false;
+}
+
+kir::Allocate* allocGlobalBufferForGridComm(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.cpp:641]: Just moved.

File: torch/csrc/jit/codegen/cuda/lower_utils.h [hunk 7]
```diff
@@ -116,13 +114,6 @@ bool derivedFromRootCAAxes(const TensorView* tv, IterDomain* axis);
 std::unordered_map<ParallelType, IterDomain*, TypeHash> getParallelDomains(
     const Val* val);
 
-// Allocate global buffer for a grid communication calls, i.e. grid reduce, grid
-// welford reduce, grid broadcast.
-kir::Allocate* allocGlobalBufferForGridComm(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.h:121]: Just moved.

File: torch/csrc/jit/codegen/cuda/lower_utils.h [hunk 8]
```diff
@@ -272,6 +223,52 @@ struct TORCH_CUDA_CU_API IterDomainDependencySorter {
...
 
+} // namespace ir_utils
+
+namespace lower_utils {
+
+bool hasBlockSync(const Expr* expr, const ThreadPredicateMap& pred_map);
+
+// Allocate global buffer for a grid communication calls, i.e. grid reduce, grid
+// welford reduce, grid broadcast.
+kir::Allocate* allocGlobalBufferForGridComm(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.h:234]: Just moved.

File: torch/csrc/jit/codegen/cuda/lower_utils.h [hunk 9]
```diff
@@ -158,49 +149,12 @@ bool isTensorScalarFillOp(const Expr* expr);
 TORCH_CUDA_CU_API std::vector<Expr*> flattenScopedExprs(
     const std::vector<Expr*>& loop_nests);
 
-//! Returns the concretized iterdomain according to
-//!  the exact compute at map.
-IterDomain* caMapExactConcreteId(IterDomain* id);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.h:163]: Removed.

File: torch/csrc/jit/codegen/cuda/lower_utils.h [hunk 10]
```diff
@@ -158,49 +149,12 @@ bool isTensorScalarFillOp(const Expr* expr);
...
 //!  in `from` and `to`.
 std::vector<Expr*> getAllSwizzlesBetween(
     std::vector<IterDomain*> from,
     std::vector<IterDomain*> to);
 
-} // namespace ir_utils
-
-namespace loop_utils {
-
-struct BasicAllocInfo {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.h:175]: Just moved.

File: torch/csrc/jit/codegen/cuda/lower_utils.h [hunk 11]
```diff
@@ -158,49 +149,12 @@ bool isTensorScalarFillOp(const Expr* expr);
...
-
-  // The allocation position relative to buffer IDs, it could be outside the
-  // compute at position if it's shared memory with a compute at inside an
-  // unswitch
-  size_t alloc_pos = 0;
-};
-
-// Fill the above allocation struct based on provided information. id_map is
-// used if we're looking at a producer tensor but loops on a consumer tensor.
-BasicAllocInfo getAllocInformation(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.h:197]: Just moved.

File: torch/csrc/jit/codegen/cuda/lower_utils.h [hunk 12]
```diff
@@ -272,6 +223,52 @@ struct TORCH_CUDA_CU_API IterDomainDependencySorter {
...
+kir::Allocate* allocGlobalBufferForGridComm(
+    Val* buffer_size,
+    DataType dtype,
+    bool zero_init);
+
+} // namespace lower_utils
+
+namespace lower_loop_utils {
+
+struct BasicAllocInfo {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.h:239]: Just moved.

File: torch/csrc/jit/codegen/cuda/lower_utils.h [hunk 13]
```diff
@@ -272,6 +223,52 @@ struct TORCH_CUDA_CU_API IterDomainDependencySorter {
...
+
+  // The allocation position relative to buffer IDs, it could be outside the
+  // compute at position if it's shared memory with a compute at inside an
+  // unswitch
+  size_t alloc_pos = 0;
+};
+
+// Fill the above allocation struct based on provided information. id_map is
+// used if we're looking at a producer tensor but loops on a consumer tensor.
+BasicAllocInfo getAllocInformation(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.h:261]: Just moved.

File: torch/csrc/jit/codegen/cuda/lower_utils.cpp [hunk 14]
```diff
@@ -734,15 +585,154 @@ std::vector<Expr*> replaceInputsInExpr(
   return ReplaceExprInput::replace(exprs, replacement_map);
 }
 
-bool isTrivialIterDomain(IterDomain* id) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.cpp:737]: Removed. Unused.

File: torch/csrc/jit/codegen/cuda/lower_utils.cpp [hunk 15]
```diff
@@ -734,15 +585,154 @@ std::vector<Expr*> replaceInputsInExpr(
...
-  auto pt = id->getParallelType();
-  return id->isReduction() || id->isBroadcast() || id->isStride() ||
-      (id->extent()->isOneInt() && id->start()->isZeroInt()) ||
-      pt == ParallelType::Vectorize ||
-      (isParallelTypeThread(pt) &&
-       !GpuLower::current()->haloInfo()->hasHaloWidth(id));
+std::vector<Expr*> getAllSwizzlesBetween(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.cpp:588]: Just moved.

File: torch/csrc/jit/codegen/cuda/lower_utils.cpp [hunk 16]
```diff
@@ -448,112 +405,6 @@ std::vector<Expr*> flattenScopedExprs(const std::vector<Expr*>& loop_nests) {
   return ExprFlattener::flatten(loop_nests);
 }
 
-IterDomain* caMapExactConcreteId(IterDomain* id) {
-  return GpuLower::current()->caMap()->getConcreteMappedID(
-      id, IdMappingMode::EXACT);
-}
-
-std::vector<Expr*> getAllSwizzlesBetween(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.cpp:456]: Just moved.

File: torch/csrc/jit/codegen/cuda/lower_utils.cpp [hunk 17]
```diff
@@ -448,112 +405,6 @@ std::vector<Expr*> flattenScopedExprs(const std::vector<Expr*>& loop_nests) {
...
-      });
-
-  return all_swizzles;
-}
-
-} // namespace ir_utils
-
-namespace loop_utils {
-
-BasicAllocInfo getAllocInformation(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.cpp:480]: Just moved.

File: torch/csrc/jit/codegen/cuda/lower_utils.cpp [hunk 18]
```diff
@@ -734,15 +585,154 @@ std::vector<Expr*> replaceInputsInExpr(
...
+      IrBuilder::create<TensorView>(buffer_domain, dtype, MemoryType::Global);
+  return IrBuilder::create<kir::Allocate>(
+      buffer_tv, buffer_tv->getMemoryType(), nullptr, zero_init);
+}
+
+} // namespace lower_utils
+
+namespace lower_loop_utils {
+
+BasicAllocInfo getAllocInformation(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.cpp:655]: Just moved.

File: torch/csrc/jit/codegen/cuda/lower_utils.h [hunk 19]
```diff
@@ -211,9 +165,6 @@ std::vector<Expr*> replaceInputsInExpr(
     const std::vector<Expr*>& exprs,
     const std::unordered_map<Val*, Val*>& replacement_map);
 
-// True if an IterDomain does not materialize a loop
-bool isTrivialIterDomain(IterDomain* id);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.h:215]: Removed, unused.

File: torch/csrc/jit/codegen/cuda/lower_utils.cpp [hunk 20]
```diff
@@ -734,15 +585,154 @@ std::vector<Expr*> replaceInputsInExpr(
...
+  const auto buffer_domain = IrBuilder::create<TensorDomain>(new_buffer_ids);
+  const auto buffer_tv =
+      IrBuilder::create<TensorView>(buffer_domain, dtype, MemoryType::Global);
+  return IrBuilder::create<kir::Allocate>(
+      buffer_tv, buffer_tv->getMemoryType(), nullptr, zero_init);
+}
+
+} // namespace lower_utils
+
+namespace lower_loop_utils {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_utils.cpp]: nit: do we need a separate namespace? Seems like just `lower_utils` would work fine.

**Review by naoyam — APPROVED** [review]
LGTM

**Pull request merged**